File: apps/macos-helper/Sources/WorkflowRecorderCore/BackendClient.swift
import Foundation

public final class BackendClient: @unchecked Sendable {
    private let config: HelperConfig
    private let decoder = JSONDecoder()
    private let encoder = JSONEncoder()

    public init(config: HelperConfig) {
        self.config = config
    }

    public func pair() async throws -> BackendDevice {
        let response: PairResponse = try await post(
            "/helper/pair",
            body: ["device_id": config.deviceID, "name": config.deviceName]
        )
        return response.device
    }

    public func pendingSessions() async throws -> [BackendSession] {
        let path = "/helper/sessions/pending?device_id=\(config.deviceID.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? config.deviceID)"
        let response: PendingSessionsResponse = try await get(path)
        return response.sessions
    }

    public func session(_ sessionID: String) async throws -> BackendSession {
        let response: SessionResponse = try await get("/helper/sessions/\(sessionID)")
        return response.session
    }

    public func confirm(sessionID: String) async throws {
        let _: SessionResponse = try await post("/helper/sessions/\(sessionID)/confirm", body: EmptyBody())
    }

    public func cancel(sessionID: String) async throws {
        let _: SessionResponse = try await post("/helper/sessions/\(sessionID)/cancel", body: EmptyBody())
    }

    public func stopped(sessionID: String) async throws {
        let _: SessionResponse = try await post("/helper/sessions/\(sessionID)/stopped", body: EmptyBody())
    }

    public func summary(sessionID: String, summary: SessionSummary) async throws {
        let _: SessionResponse = try await post("/helper/sessions/\(sessionID)/summary", body: summary)
    }

    private func get<T: Decodable>(_ path: String) async throws -> T {
        var request = URLRequest(url: config.backendBaseURL.appendingPathComponent(path))
        if path.hasPrefix("/") {
            request = URLRequest(url: URL(string: String(path.dropFirst()), relativeTo: config.backendBaseURL)!)
        }
        request.setValue(config.helperSecret, forHTTPHeaderField: "x-helper-secret")
        let (data, response) = try await URLSession.shared.data(for: request)
        try Self.validate(response: response, data: data)
        return try decoder.decode(T.self, from: data)
    }

    private func post<T: Decodable, B: Encodable>(_ path: String, body: B) async throws -> T {
        var request = URLRequest(url: URL(string: String(path.dropFirst()), relativeTo: config.backendBaseURL)!)
        request.httpMethod = "POST"
        request.setValue(config.helperSecret, forHTTPHeaderField: "x-helper-secret")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try encoder.encode(body)
        let (data, response) = try await URLSession.shared.data(for: request)
        try Self.validate(response: response, data: data)
        return try decoder.decode(T.self, from: data)
    }

    private static func validate(response: URLResponse, data: Data) throws {
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            let text = String(data: data, encoding: .utf8) ?? "<non-utf8>"
            throw HelperError.backend("Backend request failed: \(text)")
        }
    }
}

private struct EmptyBody: Encodable {}

public enum HelperError: Error, CustomStringConvertible {
    case backend(String)
    case invalidState(String)

    public var description: String {
        switch self {
        case .backend(let message), .invalidState(let message):
            return message
        }
    }
}

File: apps/dev-helper/src/backend-client.ts
import type { SessionSummary } from "@trailwise/shared";
import type { DevHelperConfig } from "./config.js";

export interface BackendSession {
  session_id: string;
  target_url: string;
  status: string;
}

export class BackendClient {
  constructor(private readonly config: DevHelperConfig) {}

  pair(): Promise<unknown> {
    return this.post("/helper/pair", {
      device_id: this.config.deviceId,
      name: this.config.deviceName
    });
  }

  async pendingSessions(): Promise<BackendSession[]> {
    const response = (await this.get(`/helper/sessions/pending?device_id=${encodeURIComponent(this.config.deviceId)}`)) as {
      sessions: BackendSession[];
    };
    return response.sessions;
  }

  async session(sessionId: string): Promise<BackendSession> {
    const response = (await this.get(`/helper/sessions/${sessionId}`)) as { session: BackendSession };
    return response.session;
  }

  confirm(sessionId: string): Promise<unknown> {
    return this.post(`/helper/sessions/${sessionId}/confirm`, {});
  }

  cancel(sessionId: string): Promise<unknown> {
    return this.post(`/helper/sessions/${sessionId}/cancel`, {});
  }

  summary(sessionId: string, summary: SessionSummary): Promise<unknown> {
    return this.post(`/helper/sessions/${sessionId}/summary`, summary);
  }

  private async get(path: string): Promise<unknown> {
    const response = await fetch(new URL(path, this.config.backendBaseUrl), {
      headers: { "x-helper-secret": this.config.helperSecret }
    });
    return this.parse(response);
  }

  private async post(path: string, body: unknown): Promise<unknown> {
    const response = await fetch(new URL(path, this.config.backendBaseUrl), {
      method: "POST",
      headers: {
        "content-type": "application/json",
        "x-helper-secret": this.config.helperSecret
      },
      body: JSON.stringify(body)
    });
    return this.parse(response);
  }

  private async parse(response: Response): Promise<unknown> {
    const text = await response.text();
    if (!response.ok) throw new Error(`Backend request failed ${response.status}: ${text}`);
    return text ? JSON.parse(text) : {};
  }
}
